Fix Doxygen warnings and spot any other ones in CI

tormodvolden opened the pull request.

tormodvolden committed:
docs: Update doxygen configuration to 1.9.8

Using doxygen 1.9.8 with the command line:
 doxygen -u doc/doxygen.cfg.in

This gets rid of the below warnings from doxygen 1.9.8:

warning: Tag 'OUTPUT_TEXT_DIRECTION' at line 102 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'HTML_TIMESTAMP' at line 1247 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'FORMULA_TRANSPARENT' at line 1552 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'LATEX_SOURCE_CODE' at line 1865 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'LATEX_TIMESTAMP' at line 1881 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'RTF_SOURCE_CODE' at line 1955 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'DOCBOOK_PROGRAMLISTING' at line 2060 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'CLASS_DIAGRAMS' at line 2250 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'DOT_FONTNAME' at line 2292 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'DOT_FONTSIZE' at line 2299 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'DOT_TRANSPARENT' at line 2545 of file 'doxygen.cfg' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"

Signed-off-by: Tormod Volden <debian.tormod@gmail.com>

tormodvolden committed:
docs: Let doxygen fail if any warnings

Signed-off-by: Tormod Volden <debian.tormod@gmail.com>

tormodvolden committed:
ci: Run doxygen on linux and macos builds

Signed-off-by: Tormod Volden <debian.tormod@gmail.com>

tormodvolden commented:
I hope there is a way to ignore doxygen.cfg warnings, but fail on other warnings. The doxygen versions are different everywhere, and there are either obsolete or unknown (future) options.

tormodvolden committed:
docs: Comment out recent clang options from doxygen.cfg.in

These options are not always available and will create unnecessary
warnings. The set values were anyway the default values.

Signed-off-by: Tormod Volden <debian.tormod@gmail.com>

tormodvolden committed:
docs: Comment out recently added doxygen options

These would create warnings when running with a slightly older
doxygen version. The set values were anyway the default values,
with the exception of TIMESTAMP: For some reason, running
doxygen -u in the earlier commit gave it the non-default YES
value.

Signed-off-by: Tormod Volden <debian.tormod@gmail.com>

tormodvolden commented:
This should go green once rebased on PR #1545. But I will also test with some other doxygen versions.

tormodvolden commented:
Note to self: The options removed with doxygen -u in commit a9e6cab all had default values, except HTML_TIMESTAMP which was set to YES. I guess HTML_TIMESTAMP was renamed or replaced with TIMESTAMP and the value was transferred.

tormodvolden commented:
I have rebased to latest master but note that there are recent documentation changes that causes doxygen errors.

mcuee commented:
```
mcuee  ~/build/libusb  21:23  git clone https://github.com/tormodvolden/libusb.git -b doxygen_warnings libusb_pr1549
Cloning into 'libusb_pr1549'...
remote: Enumerating objects: 19427, done.
remote: Counting objects: 100% (796/796), done.
remote: Compressing objects: 100% (348/348), done.
remote: Total 19427 (delta 590), reused 458 (delta 448), pack-reused 18631 (from 2)
Receiving objects: 100% (19427/19427), 6.16 MiB | 20.83 MiB/s, done.
Resolving deltas: 100% (13857/13857), done.

mcuee  …/libusb_pr1549   doxygen_warnings   21:24  ./autogen.sh 
...

mcuee  …/libusb_pr1549   doxygen_warnings     v15.2.0  21:24  make
...

mcuee  …/libusb_pr1549   doxygen_warnings ?    v15.2.0  21:24  cd doc/
./  ../  Makefile  Makefile.in  doxygen.cfg  doxygen.cfg.in  libusb.png

mcuee  …/libusb_pr1549/doc   doxygen_warnings ?  21:25  make
/usr/bin/doxygen doxygen.cfg
/home/mcuee/build/libusb/libusb_pr1549/libusb/core.c:1570: error: Found ')' without opening '(' for trailing return type '->flying_transfers_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/core.c:1848: error: Found ')' without opening '(' for trailing return type '->lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/core.c:1892: error: Found ')' without opening '(' for trailing return type '->lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/core.c:1937: error: Found ')' without opening '(' for trailing return type '->lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/core.c:2197: error: Found ')' without opening '(' for trailing return type '->lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/core.c:2233: error: Found ')' without opening '(' for trailing return type '->lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/core.c:2272: error: Found ')' without opening '(' for trailing return type '->lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/hotplug.c:295: error: Found ')' without opening '(' for trailing return type '->hotplug_cbs_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/hotplug.c:348: error: Found ')' without opening '(' for trailing return type '->hotplug_cbs_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/hotplug.c:432: error: Found ')' without opening '(' for trailing return type '->hotplug_cbs_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/hotplug.c:470: error: Found ')' without opening '(' for trailing return type '->hotplug_cbs_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:1845: error: Found ')' without opening '(' for trailing return type '->events_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:1860: error: Found ')' without opening '(' for trailing return type '->events_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:1895: error: Found ')' without opening '(' for trailing return type '->events_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:1989: error: Found ')' without opening '(' for trailing return type '->event_waiters_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:2000: error: Found ')' without opening '(' for trailing return type '->event_waiters_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:2032: error: Found ')' without opening '(' for trailing return type '->event_waiters_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:2213: error: Found ')' without opening '(' for trailing return type '->events_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:2497: error: Found ')' without opening '(' for trailing return type '->events_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:2672: error: Found ')' without opening '(' for trailing return type '->event_data_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/io.c:2819: error: Found ')' without opening '(' for trailing return type '->events_lock)...'
/home/mcuee/build/libusb/libusb_pr1549/libusb/hotplug.c:47: error: unable to resolve reference to 'libusb_hotplug_register_callback()' for \ref command
/home/mcuee/build/libusb/libusb_pr1549/libusb/hotplug.c:48: error: unable to resolve reference to 'libusb_hotplug_deregister_callback()' for \ref command
make: *** [Makefile:19: api-1.0] Error 1

```

tormodvolden commented:
> I have rebased to latest master but note that there are recent documentation changes that causes doxygen errors.

I have added all the macros from clang_thread_safety.h to the PREDEFINES, so doxygen should run without warnings now.

mcuee commented:
Yes, this is good now.

```
mcuee  …/libusb_pr1549   doxygen_warnings ?    v15.2.0  20:14  cd doc/
./  ../  Makefile  Makefile.in  doxygen.cfg  doxygen.cfg.in  libusb.png

mcuee  …/libusb_pr1549/doc   doxygen_warnings ?  20:15  make
/usr/bin/doxygen doxygen.cfg

mcuee  …/libusb_pr1549/doc   doxygen_warnings ?  20:15  ls
./  ../  Makefile  Makefile.in  api-1.0/  doxygen.cfg  doxygen.cfg.in  libusb.png  man/

mcuee  …/libusb_pr1549/doc   doxygen_warnings ?  20:15  ls api-1.0/
./                                         group__libusb__desc.html     libusb_mtasync.html                                    structlibusb__interface__association__descriptor__array.html
../                                        group__libusb__dev.html      libusb_packetoverflow.html                             structlibusb__interface__descriptor.html
annotated.html                             group__libusb__hotplug.html  menu.js                                                structlibusb__iso__packet__descriptor.html
clang__thread__safety_8h_source.html       group__libusb__lib.html      menudata.js                                            structlibusb__platform__descriptor.html
classes.html                               group__libusb__misc.html     navtree.css                                            structlibusb__pollfd.html
clipboard.js                               group__libusb__poll.html     pages.html                                             structlibusb__ss__endpoint__companion__descriptor.html
deprecated.html                            group__libusb__syncio.html   structlibusb__bos__descriptor.html                     structlibusb__ss__usb__device__capability__descriptor.html
dir_6560186179ae86ab3fbed805e7cb86e4.html  index.html                   structlibusb__bos__dev__capability__descriptor.html    structlibusb__ssplus__sublink__attribute.html
doxygen.css                                jquery.js                    structlibusb__config__descriptor.html                  structlibusb__ssplus__usb__device__capability__descriptor.html
doxygen.svg                                libusb.png                   structlibusb__container__id__descriptor.html           structlibusb__transfer.html
doxygen_crawl.html                         libusb_8h_source.html        structlibusb__control__setup.html                      structlibusb__usb__2__0__extension__descriptor.html
dynsections.js                             libusb_api.html              structlibusb__device__descriptor.html                  structlibusb__version.html
files.html                                 libusb_caveats.html          structlibusb__endpoint__descriptor.html                tabs.css
functions.html                             libusb_contexts.html         structlibusb__init__option.html                        topics.html
functions_vars.html                        libusb_hotplug.html          structlibusb__interface.html                           version__describe_8h_source.html
group__libusb__asyncio.html                libusb_io.html               structlibusb__interface__association__descriptor.html    

mcuee  …/libusb_pr1549/doc   doxygen_warnings ?  20:15  ls man
./  ../  man3/
```

mcuee commented:
Merging this as I do not want to see the warnings in the documentation build any more...

mcuee commented:
Hmm, there are problems with latest macOS build. Looks like some later macOS changes break this PR.

```
mcuee in ~/build/libusb/libusb/build/doc on master % make
/opt/homebrew/bin/doxygen doxygen.cfg
/Users/mcuee/build/libusb/libusb/libusb/descriptor.c:1833: error: unable to resolve reference to 'libusb_get_string_descriptor()' for \ref command
/Users/mcuee/build/libusb/libusb/libusb/descriptor.c:1906: error: unable to resolve reference to 'libusb_get_string_descriptor()' for \ref command
/Users/mcuee/build/libusb/libusb/libusb/libusb.h:1738: error: Member value (variable) of struct libusb_init_option is not documented.
/Users/mcuee/build/libusb/libusb/libusb/libusb.h:1736: error: Member ival (variable) of union libusb_init_option::[union].value is not documented.
/Users/mcuee/build/libusb/libusb/libusb/libusb.h:1737: error: Member log_cbval (variable) of union libusb_init_option::[union].value is not documented.
make: *** [api-1.0] Error 1
```

mcuee commented:
@tormodvolden and @Youw 

I still think it is better to fix the current macOS build and not to revert this PR. Any objections?

mcuee commented:
Info:

```
mcuee in ~/build/libusb/libusb/build/doc on master % doxygen --version
1.18.0

mcuee in ~/build/libusb/libusb/build/doc on master % brew info doxygen
==> doxygen ✔: stable 1.18.0 (bottled), HEAD
Generate documentation for several programming languages
https://www.doxygen.nl/
Installed (on request)
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/d/doxygen.rb
License: GPL-2.0-only
==> Installed Versions
doxygen ✔ 1.18.0 (7 files, 23.3MB) [Linked]
==> Dependencies
Required (2): fmt ✔, spdlog ✔
Recursive Runtime (2): all installed ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 4,160 (30 days), 19,864 (90 days), 86,991 (365 days)
install-on-request: 1,139 (30 days), 7,107 (90 days), 28,778 (365 days)
build-error: 297 (30 days)```
